### PR TITLE
GRAPH.SLOWLOG

### PR DESCRIPTION
**Pull request by swilly22**
127.0.0.1:6379> GRAPH.SLOWLOG
1) 1) Time entry was added.
    2) Command
    3) Key
    4) Query
    5) Query exec time ms.
2) ...

### CODE REVIEW

File: src/commands/cmd_context.c [hunk 1]
```diff
@@ -80,6 +82,10 @@ void CommandCtx_Free(CommandCtx *command_ctx) {
 		RedisModule_FreeThreadSafeContext(command_ctx->ctx);
 	}
 
+	// Report command to slowlog.
+	SlowLog_Add(command_ctx->command_name, command_ctx->graph_ctx->graph_name, command_ctx->query,
+				QueryCtx_GetExecutionTime());
```
Comment (jeffreylovitz) [inline src/commands/cmd_context.c]: This call should be within the `if(command_ctx->query)` condition at 89. Bulk insert commands don't populate the `query` field, nor would we want to add them to the slow log.

File: src/util/object_pool/object_pool.c [hunk 2]
```diff
@@ -79,7 +79,7 @@ static void *_ObjectPool_ReuseItem(ObjectPool *pool) {
 	return item;
 }
 
-ObjectPool *ObjectPool_New(uint64_t itemCap, uint itemSize, fpDestructor fp) {
+ObjectPool *ObjectPool_New(uint64_t itemCap, uint itemSize, void (*fp)(void *)) {
```
Comment (jeffreylovitz) [inline src/util/object_pool/object_pool.c]: Why the changes in this file?

File: src/util/object_pool/object_pool.h [hunk 3]
```diff
@@ -25,14 +25,14 @@ typedef struct {
...
-	fpDestructor destructor;    // Function pointer to a clean-up function of an item.
+	void (*destructor)(void *); // Function pointer to a clean-up function of an item.
 } ObjectPool;
 
 // Create a new ObjectPool
 // itemCap - number of items ObjectPool can hold before resizing.
 // itemSize - item size in bytes.
 // fp - destructor routine for freeing items.
-ObjectPool *ObjectPool_New(uint64_t itemCap, uint itemSize, fpDestructor fp);
+ObjectPool *ObjectPool_New(uint64_t itemCap, uint itemSize, void (*fp)(void *));
```
Comment (jeffreylovitz) [inline src/util/object_pool/object_pool.h]: Why the changes in this file?

File: src/util/thpool/thpool.c [hunk 4]
```diff
@@ -41,207 +42,222 @@ static volatile int threads_on_hold;
...
+}
+
+int thpool_get_thread_id(thpool_* thpool_p, pthread_t pthread) {
+	for(int i = 0; i < thpool_p->num_threads_alive; i++) {
+		thread *thread = thpool_p->threads[i];
+		if(thread->pthread == pthread) return thread->id;
+	}
+
+	err("thpool_get_thread_id(): Could not locate thread\n");
+	assert(0);
```
Comment (jeffreylovitz) [inline src/util/thpool/thpool.c]: If the query was issued in a MULTI block or Lua script, we can validly get to this point.

File: src/slow_log/slow_log.c [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+		introduce_item = 1;
+	} else {
+		// Not enough room, see if item should be tracked.
+		SlowLogItem *top = heap_peek(heap);
+		if(top->latency < latency) {
+			top = heap_poll(heap);
+
+			char *top_str;
+			size_t len = SlowLogItem_ToString(top, &top_str);
+			assert(raxRemove(lookup, (unsigned char *)top_str, len, NULL) == 1);
```
Comment (jeffreylovitz) [inline src/slow_log/slow_log.c]: I get a number of crashes on these calls in the multi-threaded stress tests, indicating that this code isn't thread-safe.

One sample crash occurs when the item `top` retrieved from a valid-looking heap is full of invalid memory:
```
(gdb) p *top
$5 = {
  cmd = 0x3432313000000020 <error: Cannot access memory at address 0x3432313000000020>, 
  time = 140736783809344, 
  query = 0x7fffd6017020 "\030", 
  graph_id = 0x7fffd600ea50 "\020", 
  latency = 6.9533209954723707e-310
}
```
This will, depending on contents, crash in `top_str` construction or trigger the assertion failure on trying to remove invalid inputs.

File: src/slow_log/slow_log.c [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+	rm_free(item->query);
+	rm_free(item->graph_id);
+	rm_free(item);
+}
+
+// Compares two heap record nodes.
+static int _slowlog_elem_compare(const void *A, const void *B, const void *udata) {
+	SlowLogItem *a = (SlowLogItem *)A;
+	SlowLogItem *b = (SlowLogItem *)B;
+	return -1 * (a->latency - b->latency);
```
Comment (jeffreylovitz) [inline src/slow_log/slow_log.c]: Can be written as:
`return b->latency - a->latency;`

File: src/commands/cmd_context.c (hunk 1 above)
Comment (swilly22) [reply src/commands/cmd_context.c]: 👍

File: src/util/thpool/thpool.c (hunk 4 above)
Comment (swilly22) [reply src/util/thpool/thpool.c]: 👍

File: src/slow_log/slow_log.c (hunk 6 above)
Comment (swilly22) [reply src/slow_log/slow_log.c]: 👍

File: src/slow_log/slow_log.c [hunk 7]
```diff
@@ -24,6 +24,20 @@ static int get_thread_id() {
 	return thread_id;
 }
 
+/* Redis prints doubles with up to 17 digits of precision, which captures
+ * the inaccuracy of many floating-point numbers (such as 0.1).
+ * By using the %g format and a precision of 5 significant digits, we avoid many
+ * awkward representations like RETURN 0.1 emitting "0.10000000000000001",
+ * though we're still subject to many of the typical issues with floating-point error. */
+static inline void _ReplyWithRoundedDouble(RedisModuleCtx *ctx, double d) {
```
Comment (jeffreylovitz) [inline src/slow_log/slow_log.c]: :+1: This is much cleaner.
